runtime/js/40_testing.js

```diff
@@ -823,37 +844,227 @@
     }
   }
 
-  async function runBench(bench) {
-    if (bench.ignore) {
-      return "ignored";
+  function compareMeasurements(a, b) {
+    if (a > b) return 1;
+    if (a < b) return -1;
+
+    return 0;
+  }
+
+  function benchStats(n, t, avg, min, max, all) {
+    return {
+      n,
+      min,
+      max,
+      p75: all[MathCeil(n * (75 / 100)) - 1],
+      p99: all[MathCeil(n * (99 / 100)) - 1],
+      avg: !t ? (avg / n) : MathCeil(avg / n),
+      p995: all[MathCeil(n * (99.5 / 100)) - 1],
+      p999: all[MathCeil(n * (99.9 / 100)) - 1],
+    };
+  }
+
+  function benchMeasureSync(t, fn, step) {
+    let n = 0;
+    let avg = 0;
+    let wavg = 0;
+    const all = [];
+    let min = Infinity;
+    let max = -Infinity;
+
+    warmup: {
+      let c = 0;
+      step.warmup = true;
+      let iterations = 20;
+      let budget = 10 * 1e6;
+
+      while (0 < budget || 0 < iterations--) {
+        const t1 = benchNow();
+
+        fn();
+        const t2 = benchNow() - t1;
+        if (0 > t2) {
+          iterations++;
+          continue;
+        }
+
+        c++;
+        wavg += t2;
+        budget -= t2;
+      }
+
+      wavg /= c;
+      break warmup;
+    }
+
+    measure: {
+      step.warmup = false;
+
+      if (wavg > 10_000) {
+        let iterations = 10;
+        let budget = t * 1e6;
+
+        while (0 < budget || 0 < iterations--) {
+          const t1 = benchNow();
+
+          fn();
+          const t2 = benchNow() - t1;
+          if (0 > t2) {
+            iterations++;
+            continue;
+          }
+
+          n++;
+          avg += t2;
+          budget -= t2;
+          all.push(t2);
+          if (t2 < min) min = t2;
+          if (t2 > max) max = t2;
+        }
+      } else {
+        let iterations = 10;
+        let budget = t * 1e6;
+
+        while (0 < budget || 0 < iterations--) {
+          const t1 = benchNow();
+          for (let c = 0; c < 1e4; c++) fn();
+          const t2 = (benchNow() - t1) / 1e4;
+          if (0 > t2) {
+            iterations++;
+            continue;
+          }
+
+          n++;
+          avg += t2;
+          all.push(t2);
+          budget -= t2 * 1e4;
+          if (t2 < min) min = t2;
+          if (t2 > max) max = t2;
+        }
+      }
+
+      break measure;
     }
 
+    all.sort(compareMeasurements);
+    return benchStats(n, wavg > 10_000, avg, min, max, all);
+  }
+
+  async function benchMeasureAsync(t, fn, step) {
```

Comment:
There's a lot of duplicate code between `benchMeasureSync` and `benchMeasureAsync`. I think we only need to duplicate the measurement code though?

```
const t1 = benchNow();
await fn();
const elapsed = benchNow() - t1;
```

...and...

```
const t1 = benchNow();
fn();
const elapsed = benchNow() - t1;
```